src/lib/statistics/statistics_objects/top_k_uniform_distribution_histogram.cpp

```diff
@@ -0,0 +1,154 @@
+#include "top_k_uniform_distribution_histogram.hpp"
+
+#include <cmath>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+#include <iterator>
+#include <algorithm>
+
+#include <tsl/robin_map.h>  // NOLINT
+
+#include "generic_histogram.hpp"
+#include "resolve_type.hpp"
+#include "storage/segment_iterate.hpp"
+
+#include "expression/evaluation/like_matcher.hpp"
+#include "generic_histogram.hpp"
+#include "generic_histogram_builder.hpp"
+#include "lossy_cast.hpp"
+#include "resolve_type.hpp"
+#include "statistics/statistics_objects/abstract_statistics_object.hpp"
+#include "storage/create_iterable_from_segment.hpp"
+#include "storage/segment_iterate.hpp"
+#include "abstract_histogram.hpp"
+#include "equal_distinct_count_histogram.hpp"
+#include "equal_distinct_count_histogram.cpp"
+
+
+namespace opossum {
+
+template <typename T>
+std::shared_ptr<GenericHistogram<T>> TopKUniformDistributionHistogram<T>::from_column(
+    const Table& table, const ColumnID column_id, const BinID max_bin_count, const HistogramDomain<T>& domain) {
+  Assert(max_bin_count > 0, "max_bin_count must be greater than zero ");
+
+  auto value_distribution = value_distribution_from_column(table, column_id, domain);
+
+  if (value_distribution.empty()) {
+    return nullptr;
+  }
+
+  // If the column holds less than K distinct values use the distinct count as TOP_K instead
+
+  const auto k = std::min(TOP_K_DEFAULT, value_distribution.size());
+
+  // Get the first top k values and save them into vectors
+  std::vector<T> top_k_names(k);
+  std::vector<HistogramCountType> top_k_counts(k);
+
+  // Sort values by occurrence count
+  auto sorted_count_values = value_distribution;
+  std::sort(sorted_count_values.begin(), sorted_count_values.end(),
+            [&](const auto& l, const auto& r) { return l.second > r.second; });
+
+  // Sort TOP_K values with highest occurrence count lexicographically.
+  // We later use this for more performant range predicate evaluation.
+  std::sort(sorted_count_values.begin(), sorted_count_values.begin() + k,
+            [&](const auto& l, const auto& r) { return l.first < r.first; });
+
+  if (!sorted_count_values.empty()) {
+    for(auto i = 0u; i < k; i++) {
+      top_k_names[i] = sorted_count_values[i].first;
+      top_k_counts[i] = sorted_count_values[i].second;
+    }
+  }
+
+  // Remove TOP_K values from value distribution
+  for (auto i = 0u; i < k; i++) {
+    auto it = remove(value_distribution.begin(), value_distribution.end(), std::make_pair(top_k_names[i], top_k_counts[i]));
+    value_distribution.erase(it, value_distribution.end());
+  }
```

Comment:
This is quite expensive. Is it possible to make it cheaper? Save indices of top_k values, swap them to the end and then drop them or something like this?